azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -160,38 +198,142 @@ def __get_agent_family_manifests(self, goal_state):
+
+            # verify if agent update is allowed
+            if not updater.should_update_agent(goal_state):
+                return
+            updater.log_new_agent_update_message()
+            updater.purge_extra_agents_from_disk()
+            agent = updater.download_and_get_new_agent(self._protocol, goal_state)
+            if agent.is_blacklisted or not agent.is_downloaded:
+                msg = "Downloaded agent version is in bad state : {0} , skipping agent update".format(
+                    str(agent.version))
+                raise AgentUpdateError(msg)
+            updater.proceed_with_update()
+
+        except Exception as err:
+            if isinstance(err, AgentUpgradeExitException):
+                raise err
+            elif isinstance(err, AgentUpdateError):
+                error_msg = ustr(err)
+            else:
+                error_msg = "Unable to update Agent: {0}".format(textutil.format_exception(err))
+            logger.warn(error_msg)
+            add_event(op=WALAEventOperation.AgentUpgrade, is_success=False, message=error_msg, log_event=False)
+            self.update_state.last_attempted_update_error_msg = error_msg
+
+    def get_vmagent_update_status(self):
+        """
+        This function gets the VMAgent update status as per the last attempted update.
+        Returns: None if fail to report or update never attempted with rsm version specified in GS
+        """
+        try:
+            if conf.get_enable_ga_versioning() and self._is_vm_enabled_for_rsm_upgrades and self._is_version_from_rsm:
+                if not self.update_state.last_attempted_update_error_msg:
+                    status = VMAgentUpdateStatuses.Success
+                    code = 0
+                else:
+                    status = VMAgentUpdateStatuses.Error
+                    code = 1
+                return VMAgentUpdateStatus(expected_version=str(self.update_state.last_attempted_update_version), status=status, code=code, message=self.update_state.last_attempted_update_error_msg)
+        except Exception as err:
+            msg = "Unable to report agent update status: {0}".format(textutil.format_exception(err))
+            logger.warn(msg)
+            add_event(op=WALAEventOperation.AgentUpgrade, is_success=False, message=msg, log_event=True)
+        return None
 
-    def __download_and_get_agent(self, goal_state, agent_family, agent_manifest, requested_version):
+
+class GAVersionUpdater(object):
+
+    def __init__(self, gs_id, agent_family, agent_manifest, version, update_state):
+        self._gs_id = gs_id
+        self._agent_family = agent_family
+        self._agent_manifest = agent_manifest
+        self._version = version
+        self._update_state = update_state
+
+    def should_update_agent(self, goal_state):
         """
-        This function downloads the new agent(requested version) and returns the downloaded version.
+        RSM version update:
```

Comment:
These comments should probably be in the derived classes to make the code easier to read instead of having to scroll up and down to see the comments